src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,370 @@
+		ereport(LOG, errmsg("%ld", average));
+	}
+
+	SPI_finish();
+
+	if (rowCount > 0)
+	{
+		list_sort(SplitPoints, list_int_cmp);
+	}
+	else
+	{
+		if (shardinfo->shardminvalue <= average)
+		{
+			SplitPoints = lappend_int(SplitPoints, average);
+		}
+	}
+
+	return SplitPoints;
+}
+
+
+/*
+ * This function calculates the split points of the shard to
+ * split and then executes the background job for the shard split.
+ */
+void
+ScheduleShardSplit(ShardInfo shardinfo , char* shardSplitMode)
+{
+	List *SplitPoints = FindShardSplitPoints(shardinfo);
+	if (list_length(SplitPoints) > 0)
+	{
+		// ErrorOnConcurrentOperation();
+		int64 jobId = CreateBackgroundJob("Automatic Shard Split",
+										  "Split using SplitPoints List");
+		ereport(LOG,errmsg("%s",GetShardSplitQuery(shardinfo,SplitPoints,shardSplitMode)->data));					
+		ExecuteSplitBackgroundJob(jobId, shardinfo, SplitPoints , shardSplitMode);
+	}
+	else
+	{
+		ereport(LOG, errmsg("No Splitpoints for shard split"));
+	}
+}
+
+
+/*
+ * citus_auto_shard_split_start finds shards whose colocation group has total size greater than threshold
+ * and from that extracts the shard with the maximum size . It isolates the shards if the shard has common
+ * tenant values otherwise it 2 way splits it on the basis of average hash value.
+ *
+ */
+Datum
+citus_auto_shard_split_start(PG_FUNCTION_ARGS)
+{
+	StringInfo query = makeStringInfo();
+
+	/* This query is written to group the shards on the basis of colocation id and shardminvalue and get the groups whose sum of shardsize
+	 * are greater than a threshold and than extract the shard in them which has the maximum size. So for that first pg_dist_shard and citus_shards are joined followed by the joining of pg_dist_node
+	 * and citus_shards and finally joined by the table obtained by the grouping of colocation id and shardminvalue and shardsize exceeding the threshold.*/
+
+	appendStringInfo(
+		query,
+		" SELECT cs.shardid,pd.shardminvalue,pd.shardmaxvalue,cs.shard_size,pn.nodeid,ct.distribution_column,ct.table_name,cs.shard_name,(SELECT relname FROM pg_class WHERE oid = (ct.table_name::regclass)::oid) "
+		" FROM pg_catalog.pg_dist_shard pd JOIN pg_catalog.citus_shards cs ON pd.shardid = cs.shardid JOIN pg_catalog.pg_dist_node pn ON cs.nodename = pn.nodename AND cs.nodeport= pn.nodeport"
+		" JOIN"
+		" ( select shardid , max_size from (SELECT distinct first_value(shardid) OVER w as shardid, sum(shard_size) OVER (PARTITION BY colocation_id, shardminvalue) as total_sum, max(shard_size) OVER w as max_size"
+		" FROM citus_shards cs JOIN pg_dist_shard ps USING(shardid)"
+		" WINDOW w AS (PARTITION BY colocation_id, shardminvalue ORDER BY shard_size DESC) )as t where total_sum >= %ld )"
+		" AS max_sizes ON cs.shardid=max_sizes.shardid AND cs.shard_size = max_sizes.max_size JOIN citus_tables ct ON cs.table_name = ct.table_name AND pd.shardminvalue <> pd.shardmaxvalue AND pd.shardminvalue <> ''",
+		MaxShardSize*1024
```

Comment:
you might want to use uint64 here, because otherwise it might overflow

Reply:
used uint64